File: verify.py
#!/usr/bin/env python3
"""RiverAlpha chain verifier — stdlib only. Recomputes every hash from raw data.

Usage: python verify.py [path/to/rounds.jsonl]
"""
import hashlib
import json
import sys


def H(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


GENESIS = H("hangang-pj/genesis")


def canon(o) -> str:
    return json.dumps(o, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def merkle(leaf_hashes: list[str]) -> str:
    if not leaf_hashes:
        return H("")
    level = [H("00" + leaf) for leaf in leaf_hashes]
    while len(level) > 1:
        nxt = []
        for i in range(0, len(level), 2):
            a = level[i]
            b = level[i + 1] if i + 1 < len(level) else a
            nxt.append(H("01" + a + b))
        level = nxt
    return level[0]


def main() -> int:
    path = sys.argv[1] if len(sys.argv) > 1 else "data/log/crypto/rounds.jsonl"
    chain = GENESIS
    n = 0
    with open(path, encoding="utf-8") as f:
        for line in f:
            if not line.strip():
                continue
            r = json.loads(line)
            preds = sorted(r["predictions"], key=lambda p: p["participantId"])
            leaves = []
            for p in preds:
                payload = {
                    "participantId": p["participantId"],
                    "roundId": r["roundId"],
                    "signal": p["signal"],
                }
                h = H(canon(payload))
                if "payloadHash" in p and h != p["payloadHash"]:
                    print(f"FAIL {r['roundId']}: payload hash mismatch ({p['participantId']})")
                    return 1
                leaves.append(h)
            root = merkle(leaves)
            if root != r["merkleRoot"]:
                print(f"FAIL {r['roundId']}: merkle root mismatch")
                return 1
            if r["prevChainHash"] != chain:
                print(f"FAIL {r['roundId']}: prev chain hash mismatch")
                return 1
            chain = H("02" + chain + root)
            if chain != r["chainHash"]:
                print(f"FAIL {r['roundId']}: chain hash mismatch")
                return 1
            n += 1
            print(f"OK   {r['roundId']}  chain={chain[:16]}…")
    print(f"\nchain intact: {n} round(s) verified")
    print(f"latest chain hash: {chain}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: verify.mjs
// 일회용 교차검증: 엔진 코드 없이 README 스펙만으로 체인 재계산 (verify.py와 동일 로직)
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";

const H = (s) => createHash("sha256").update(s, "utf8").digest("hex");
const canon = (o) => {
  if (o === null || typeof o !== "object") return JSON.stringify(o);
  if (Array.isArray(o)) return "[" + o.map(canon).join(",") + "]";
  return "{" + Object.keys(o).sort().map((k) => JSON.stringify(k) + ":" + canon(o[k])).join(",") + "}";
};
const merkle = (leaves) => {
  if (!leaves.length) return H("");
  let lv = leaves.map((l) => H("00" + l));
  while (lv.length > 1) {
    const nx = [];
    for (let i = 0; i < lv.length; i += 2) nx.push(H("01" + lv[i] + (lv[i + 1] ?? lv[i])));
    lv = nx;
  }
  return lv[0];
};

let chain = H("hangang-pj/genesis");
let n = 0;
const path = process.argv[2] ?? "data/log/crypto/rounds.jsonl";
for (const line of readFileSync(path, "utf8").trim().split("\n")) {
  const r = JSON.parse(line);
  const leaves = [...r.predictions]
    .sort((a, b) => (a.participantId < b.participantId ? -1 : 1))
    .map((p) => {
      const payload = { participantId: p.participantId, roundId: r.roundId, signal: p.signal };
      if (p.evidenceHash) {
        // 2026-07-09~: 감사 증거(모델·파라미터·promptHash·rationale·원응답)도 leaf에 바인딩
        const ev = H(canon({
          modelRequested: p.modelRequested,
          modelUsed: p.modelUsed,
          params: p.params,
          promptHash: p.promptHash,
          rationale: p.rationale,
          raw: p.raw,
        }));
        if (ev !== p.evidenceHash) throw new Error(`${r.roundId}: evidence mismatch (${p.participantId})`);
        payload.evidenceHash = p.evidenceHash;
      }
      const h = H(canon(payload));
      if (p.payloadHash && h !== p.payloadHash) throw new Error(`${r.roundId}: payload mismatch (${p.participantId})`);
      return h;
    });
  const root = merkle(leaves);
  if (root !== r.merkleRoot) throw new Error(`${r.roundId}: merkle mismatch`);
  if (r.prevChainHash !== chain) throw new Error(`${r.roundId}: prev mismatch`);
  chain = H("02" + chain + root);
  if (chain !== r.chainHash) throw new Error(`${r.roundId}: chain mismatch`);
  n++;
}
console.log(`independent crosscheck OK: ${n} round(s), chain=${chain}`);
